File: Exercise_3.cpp
// Time Complexity                              :  O(N)
// Space Complexity                             :  O(1)
// Did this code successfully run on Leetcode   :  Yes
// Any problem you faced while coding this      :  No

#include <bits/stdc++.h> 
using namespace std;  


// https://leetcode.com/problems/search-a-2d-matrix-ii/
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        
        int n = matrix.size();
        int m = matrix[0].size();
        if(n == 0) return false;
        int i=n-1;int j=0;
        
        while(i>=0 and i<n and j>=0 and j<m) {
            if(target == matrix[i][j]) return true;
            else if (target > matrix[i][j]) {
                j++;
            } else {
                i--;
            }
        }
        
        return false;
        
    }
};

File: Exercise_3.js
// Time Complexity : O(m + n)
// Space Complexity : O(1)

// search 2d matrix
var searchMatrix = function (matrix, target) {
    let m = matrix.length;
    let n = matrix[0].length;
    let i = 0;
    let j = n - 1;

    while (i < m && j >= 0) {
        if (matrix[i][j] === target) {
            return true;
        }
        if (matrix[i][j] < target) {
            i += 1;
        } else {
            j -= 1;
        }
    }
    return false;
}